testutil/fs_hook.go

```diff
@@ -0,0 +1,210 @@
+// +build linux darwin
+
+// Copyright 2018 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testutil
+
+import (
+	"fmt"
+	"path/filepath"
+	"time"
+
+	"github.com/hanwen/go-fuse/fuse"
+	"github.com/hanwen/go-fuse/fuse/nodefs"
+	"github.com/hanwen/go-fuse/fuse/pathfs"
+)
+
+type HookFs struct {
+	Original   string
+	Mountpoint string
+	FsName     string
+	fs         pathfs.FileSystem
+	hook       Hook
```

Comment:
Because there will be many Hook type in the future.

For example:
```
HookOnOpen //called on open
HookOnRead // called on read
HookOnWrite //called on write
HookOnMkdir //called on mkdir
HookOnRmdir 
HookOnOpenDir
HookOnFsync
...
```
They are also implements Hook. As i said, in this pull request, I just implement fuse framework and i will add all fuse support in another pull request.

Reply:
ok, noted, but lets try with an exact type and will think about the other ones with the other PR.

Reply:
It is an obviously interface, do you think it is very necessary to update Hook to HookOnRename?

Reply:
Yes I think typecasting is error prone and should use it at last resort.

Reply:
I have this idea using composition:

```

type Hook interface {
	PreRename(oldPatgh string, newPath string) (hooked bool, err error)
	PostRename(oldPatgh string, newPath string) (hooked bool, err error)
	Chmod(name string, mode uint32, context *fuse.Context) fuse.Status
	.
	.
	. // All other funcs from the fuse server.
}

type TestRenameHook struct {
	testutil.EmptyHook // Add the Empty hook so the that this struct implements the hook interface.
}

// These will take precedence over the `testutil.EmptyHook`
func (h TestRenameHook) PreRename(oldPatgh string, newPath string) (hooked bool, err error) {
	fmt.Printf("renamed file from %s to %s \n", oldPatgh, newPath)
	return true, syscall.EIO
}
func (h TestRenameHook) PostRename(oldPatgh string, newPath string) (hooked bool, err error) {
	return false, nil
}


type EmptyHook struct{}

func (h EmptyHook) PreRename(oldPatgh string, newPath string) (hooked bool, err error) {
	return false, nil
}
func (h EmptyHook) PostRename(oldPatgh string, newPath string) (hooked bool, err error) {
	return false, nil
}
func (h EmptyHook) Chmod(name string, mode uint32, context *fuse.Context) fuse.Status { return fuse.Status(0) }
.
.
.
// All other funcs from the Hook interface.

```